File: mycelix-workspace/mycelix-supplychain/ts/adapters/csv/src/ingest.ts
#!/usr/bin/env node

// Copyright (C) 2024-2026 Tristan Stoltz / Luminous Dynamics
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial licensing: see COMMERCIAL_LICENSE.md at repository root
/**
 * CSV Adapter for Mycelix Supply Chain
 *
 * Reads CSV files with supply chain events and ingests them via the API
 */

import { SupplyChainClient, SupplyEventVC, EventType } from '@mycelix/supplychain-sdk';
import { parse } from 'csv-parse';
import { createReadStream } from 'fs';
import { program } from 'commander';

interface CsvRow {
  eventType: EventType;
  issuer: string;
  productId: string;
  batchId: string;
  prevBatchIds?: string;
  quantity: string;
  unit: string;
  facilityId: string;
  facilityName: string;
  timestamp?: string;
}

program
  .name('csv-adapter')
  .description('Ingest supply chain events from CSV files')
  .requiredOption('-f, --file <path>', 'CSV file to ingest')
  .option('-u, --url <url>', 'API base URL', 'http://localhost:8080')
  .option('-d, --dry-run', 'Parse CSV without sending to API')
  .parse();

const options = program.opts();

async function ingestCsv(filePath: string, apiUrl: string, dryRun = false) {
  const client = new SupplyChainClient({ baseUrl: apiUrl });

  // Test connection
  if (!dryRun) {
    try {
      const health = await client.health();
      console.log(`Connected to API (version ${health.version})`);
    } catch (error) {
      console.error('Failed to connect to API:', error);
      process.exit(1);
    }
  }

  const parser = createReadStream(filePath).pipe(
    parse({
      columns: true,
      skip_empty_lines: true,
      trim: true,
    })
  );

  let processed = 0;
  let errors = 0;

  for await (const row of parser as AsyncIterable<CsvRow>) {
    try {
      const event: SupplyEventVC = {
        '@context': [
          'https://www.w3.org/2018/credentials/v1',
          'https://mycelix.org/contexts/supply-chain/v1',
        ],
        type: ['VerifiableCredential', 'SupplyChainEvent'],
        issuer: row.issuer,
        issuanceDate: new Date().toISOString(),
        credentialSubject: {
          eventType: row.eventType,
          productId: row.productId,
          batchId: row.batchId,
          prevBatchIds: row.prevBatchIds ? row.prevBatchIds.split(',').map(s => s.trim()) : undefined,
          quantity: parseFloat(row.quantity),
          unit: row.unit,
          facility: {
            id: row.facilityId,
            name: row.facilityName,
          },
          timestamp: row.timestamp || new Date().toISOString(),
        },
      };

      if (dryRun) {
        console.log(`[DRY RUN] Would ingest: ${row.eventType} for batch ${row.batchId}`);
      } else {
        const result = await client.ingestEvent(event);
        console.log(`✓ Ingested ${row.eventType} for batch ${row.batchId} → claim ${result.claim_id}`);
      }

      processed++;
    } catch (error) {
      console.error(`✗ Failed to ingest row:`, row, error);
      errors++;
    }
  }

  console.log(`\nSummary: ${processed} processed, ${errors} errors`);
}

ingestCsv(options.file, options.url, options.dryRun).catch(console.error);

File: mycelix-workspace/mycelix-supplychain/adapters/python/ingest.py
#!/usr/bin/env python3

# Copyright (C) 2024-2026 Tristan Stoltz / Luminous Dynamics
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial licensing: see COMMERCIAL_LICENSE.md at repository root

# Copyright (C) 2024-2026 Tristan Stoltz / Luminous Dynamics
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial licensing: see COMMERCIAL_LICENSE.md at repository root
"""
Python Adapter for Mycelix Supply Chain

Ingest supply chain events from various data sources (CSV, Excel, ERP exports)
"""

import argparse
import json
import sys
from datetime import datetime
from typing import Any, Dict, List

import pandas as pd
import requests
from dotenv import load_dotenv


class SupplyChainClient:
    """Client for Mycelix Supply Chain API"""

    def __init__(self, base_url: str):
        self.base_url = base_url.rstrip("/")
        self.session = requests.Session()
        self.session.headers.update({"Content-Type": "application/json"})

    def health(self) -> Dict[str, Any]:
        """Check API health"""
        response = self.session.get(f"{self.base_url}/health")
        response.raise_for_status()
        return response.json()

    def ingest_event(self, event: Dict[str, Any]) -> Dict[str, Any]:
        """Ingest a supply chain event"""
        response = self.session.post(f"{self.base_url}/v1/events", json=event)
        response.raise_for_status()
        return response.json()


def csv_to_events(csv_path: str) -> List[Dict[str, Any]]:
    """Convert CSV to supply chain events"""
    df = pd.read_csv(csv_path)

    events = []
    for _, row in df.iterrows():
        event = {
            "@context": [
                "https://www.w3.org/2018/credentials/v1",
                "https://mycelix.org/contexts/supply-chain/v1",
            ],
            "type": ["VerifiableCredential", "SupplyChainEvent"],
            "issuer": row["issuer"],
            "issuanceDate": datetime.utcnow().isoformat() + "Z",
            "credentialSubject": {
                "eventType": row["eventType"],
                "productId": row["productId"],
                "batchId": row["batchId"],
                "quantity": float(row["quantity"]),
                "unit": row["unit"],
                "facility": {
                    "id": row["facilityId"],
                    "name": row["facilityName"],
                },
                "timestamp": row.get("timestamp", datetime.utcnow().isoformat() + "Z"),
            },
        }

        # Optional fields
        if "prevBatchIds" in row and pd.notna(row["prevBatchIds"]):
            event["credentialSubject"]["prevBatchIds"] = [
                b.strip() for b in row["prevBatchIds"].split(",")
            ]

        events.append(event)

    return events


def main():
    parser = argparse.ArgumentParser(description="Ingest supply chain events from CSV")
    parser.add_argument("-f", "--file", required=True, help="CSV file to ingest")
    parser.add_argument(
        "-u", "--url", default="http://localhost:8080", help="API base URL"
    )
    parser.add_argument(
        "--dry-run", action="store_true", help="Parse without sending to API"
    )

    args = parser.parse_args()

    load_dotenv()

    client = SupplyChainClient(args.url)

    # Test connection
    if not args.dry_run:
        try:
            health = client.health()
            print(f"Connected to API (version {health['version']})")
        except Exception as e:
            print(f"Failed to connect to API: {e}", file=sys.stderr)
            sys.exit(1)

    # Parse CSV
    try:
        events = csv_to_events(args.file)
        print(f"Parsed {len(events)} events from {args.file}")
    except Exception as e:
        print(f"Failed to parse CSV: {e}", file=sys.stderr)
        sys.exit(1)

    # Ingest events
    processed = 0
    errors = 0

    for event in events:
        try:
            if args.dry_run:
                print(
                    f"[DRY RUN] Would ingest: {event['credentialSubject']['eventType']} "
                    f"for batch {event['credentialSubject']['batchId']}"
                )
            else:
                result = client.ingest_event(event)
                print(
                    f"✓ Ingested {event['credentialSubject']['eventType']} "
                    f"for batch {event['credentialSubject']['batchId']} "
                    f"→ claim {result['claim_id']}"
                )
            processed += 1
        except Exception as e:
            print(f"✗ Failed to ingest event: {e}", file=sys.stderr)
            errors += 1

    print(f"\nSummary: {processed} processed, {errors} errors")


if __name__ == "__main__":
    main()
